torch/csrc/utils/tensor_new.cpp

```diff
@@ -1710,7 +1710,9 @@ Tensor asarray(
       if (wrong_device || wrong_dtype) {
         tensor = tensor.to(
             device.value_or(tensor.device()),
-            dtype.value_or(tensor.scalar_type()));
+            dtype.value_or(tensor.scalar_type()),
+            /*non_blocking=*/false,
+            /*copy=*/true);
```

Comment:
I think this creates another issue:

```python
>>> a = torch.arange(5, device="cuda")
>>> b = torch.asarray(a, device="cuda")  # a and b may be aligned.
>>> a.data_ptr() == b.data_ptr()
True
```

So, instead of adding the `copy` argument there, I think it's better to normalize the optional device argument by giving an index to it (getting from [the current device](https://pytorch.org/docs/stable/generated/torch.cuda.current_device.html#torch.cuda.current_device)). And, then, compare it against the `tensor.device()`.

Reply:
Sorry, I thought `asarray` would always return a new tensor. Could you please help me and see if I can modify it like this? Of course, I can modify it in the way you suggested, if the current modification method still has problems or is not the optimal way.

Reply:
Yeah, I think that the current code works.